File: firebase/functions/src/async_ttl_cache.ts
export class AsyncTtlCache<T> {
  constructor(
    private readonly ttlMilliseconds: number,
    private readonly now: () => number = Date.now,
  ) {}

  private value?: T;
  private expiresAt = 0;
  private pending?: Promise<T>;

  get(loader: () => Promise<T>): Promise<T> {
    if (this.value !== undefined && this.now() < this.expiresAt) {
      return Promise.resolve(this.value);
    }
    if (this.pending) return this.pending;

    const request = loader();
    this.pending = request;
    return request.then((value) => {
      this.value = value;
      this.expiresAt = this.now() + this.ttlMilliseconds;
      return value;
    }).finally(() => {
      this.pending = undefined;
    });
  }
}

File: apps/mobile/lib/src/core/data/async_ttl_cache.dart
class AsyncTtlCache<T> {
  AsyncTtlCache({DateTime Function()? now}) : _now = now ?? DateTime.now;

  final DateTime Function() _now;
  T? _value;
  DateTime? _expiresAt;
  Future<T>? _pending;
  int _requestVersion = 0;

  Future<T> get(
    Duration duration,
    Future<T> Function() loader, {
    bool forceRefresh = false,
  }) {
    final value = _value;
    final expiresAt = _expiresAt;
    if (!forceRefresh &&
        value != null &&
        expiresAt != null &&
        _now().isBefore(expiresAt)) {
      return Future.value(value);
    }

    final pending = _pending;
    if (!forceRefresh && pending != null) return pending;

    final requestVersion = ++_requestVersion;
    final request = loader();
    _pending = request;
    return request
        .then((result) {
          if (requestVersion == _requestVersion) {
            _value = result;
            _expiresAt = _now().add(duration);
          }
          return result;
        })
        .whenComplete(() {
          if (requestVersion == _requestVersion) _pending = null;
        });
  }

  void clear() {
    _value = null;
    _expiresAt = null;
    _pending = null;
    _requestVersion++;
  }
}
